cornell-dti/course-plan
# Initial Self-Check frontend implementation

### Summary 






Started the frontend implementation of the self-check requirements on the sidebar as per this [notion](https://www.notion.so/82bf355c4b7b49f99691eb9f3afb203c?v=5da9d646f3ef424bbceb1ca0e8e00cf8&p=7a3e7595daa74cc992487adc2d679e31) item.

- [x] Added a dropdown to self-checks populated with courses on the user's schedule
- [x] Selecting a course displays a completed requirement
- [x] Resetting a requirement has it disappear from the plan and from self check
- [x] Added a dropdown item to add courses from a new condensed add modal
- [x] Removed semester select from the old add modal entirely
- [x] Factored out a bunch of shared functions

![image](https://user-images.githubusercontent.com/25535093/108611901-31fcf880-73b1-11eb-8970-48635d5d848f.png)



Remaining TODOs in the sprint over the next 2 weeks.

- [ ] Update dropdown styling. I copied the toggleable requirements dropdown but both could use an improvement to look more like the Figma designs. Also need to add some more text ("Add Course") and margins.
- [ ] Only show courses that are not mapped to requirements in the dropdown
- [ ] Map courses to self-check requirements so they actually fulfill them
- [ ] Store requirement choices on firebase so they last after refresh (see #304)
- [ ] Limit the self-check dropdown from appearing after a course/credit limit for self-checks, need data for this first @einc 
- [ ] Resolve codebase changes, see questions in Notes section

### Test Plan 

1. Add an existing course to a self-check requirement, ensure that it shows up as completed
2. Add a new course from the self-check add modal
3. Reset both types of courses and ensure they are removed from the plan and sidebar
4. Ensure other subreqs (like toggleable requirements) still work as expected
5. Ensure adding from the original add modal still works as expected

### Notes 

Have a few changes I made to the codebase that I am unsure about...
- I duplicated a lot of adding course code to requirements sidebar files. I noticed later that there was already some unused code to call the add modal (the $parent.$parent etc). Should I use that existing $parent method, keep the code I duplicated and remove the unused code, or try to make some of the code I copied global so it is not duplicated anywhere?
- Design question: If either of the last two options for the above is picked, then I made some changes to the existing toggleable dropdown styling to not show up above the sidebar to get the add modal to work properly. Can I leave it like it is now, or should I try and find a way to change it back.
- I also duplicated some dropdown code & styling from toggleable requirements. I decided not to try to make it more accessible or anything since the code already existed, and we talked about maybe trying to make a dropdown component. Is this ok or should I try something different?
- @SamChou19815 I noticed that courses from semesters are a different type (FirestoreSemesterCourse) vs the ones mapped to requirements (CourseTaken). I ended up updating CompletedSubReqCourse to handle both types, but I was not sure if I was doing something wrong or should somehow be using the same type as before.
